File: src/main/java/net/onelitefeather/stardust/service/SyncFrogService.java
package net.onelitefeather.stardust.service;

import io.papermc.paper.event.player.PrePlayerAttackEntityEvent;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.onelitefeather.stardust.StardustPlugin;
import org.bukkit.*;
import org.bukkit.command.CommandSender;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Frog;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.util.StringUtil;
import org.incendo.cloud.annotation.specifier.Quoted;
import org.incendo.cloud.annotations.Argument;
import org.incendo.cloud.annotations.Command;
import org.incendo.cloud.annotations.CommandDescription;
import org.incendo.cloud.annotations.Permission;
import org.incendo.cloud.annotations.suggestion.Suggestions;
import org.incendo.cloud.context.CommandContext;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class SyncFrogService implements Listener {

    private static final String FROG_BUCKET_NAME = "Frog Bucket";
    private static final Frog.Variant DEFAULT_VARIANT = Frog.Variant.TEMPERATE;

    private final StardustPlugin plugin;
    private final NamespacedKey frogNamespacedKey;
    private final NamespacedKey frogVariantKey;
    private final NamespacedKey frogCustomNameKey;

    public SyncFrogService(StardustPlugin plugin) {
        this.plugin = plugin;
        this.frogNamespacedKey = new NamespacedKey(plugin, "frog_data_key");
        this.frogVariantKey = new NamespacedKey(plugin, "frog_variant_key");
        this.frogCustomNameKey = new NamespacedKey(plugin, "frog_custom_name_key");
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    @Command("frogbucket <customName> <variant> <amount>")
    @CommandDescription("Gives you a Frog bucket with the given amount!")
    @Permission("stardust.command.frogbucket")
    public void executeCommand(Player player,
                               @Quoted @Argument(value = "customName") String customName,
                               @Argument(value = "variant", suggestions = "frogVariants") String variant,
                               @Argument(value = "amount") int amount) {
        addFrogBucketToPlayer(
                player,
                amount,
                MiniMessage.miniMessage().deserialize(customName),
                frogVariant(variant),
                false
        );
    }

    @EventHandler
    public void handlePlayerInteract(PlayerInteractEvent event) {

        var player = event.getPlayer();

        var itemStack = event.getItem();
        if(itemStack == null) return;

        var clickedBlock = event.getClickedBlock();
        if (clickedBlock == null) return;

        if (event.getAction() != Action.RIGHT_CLICK_BLOCK) return;

        if (!isFrogInBucket(itemStack)) return;
        if (deserializeFrogData(itemStack, clickedBlock.getLocation().add(0.0, 1.0, 0.0))) {
            removeFrogBucket(player, player.getInventory().getItemInMainHand(), true);
            player.sendMessage(Component.translatable("frog-bucket-spawn-success").arguments(plugin.getPrefix()));
        } else {
            player.sendMessage(Component.translatable("frog-cannot-be-spawned").arguments(plugin.getPrefix()));
        }
    }

    @EventHandler
    public void handlePlayerPreAttack(PrePlayerAttackEntityEvent event) {

        var player = event.getPlayer();
        var entity = event.getAttacked();
        var itemInHand = player.getInventory().getItemInMainHand();

        if (isFrogInBucket(itemInHand)) return;

        if (entity instanceof Frog frog && itemInHand.getType() == Material.BUCKET) {

            var customName = entity.customName() != null ? entity.customName() : MiniMessage.miniMessage().deserialize(entity.getName());

            addFrogBucketToPlayer(
                    player,
                    1,
                    customName,
                    frog.getVariant(),
                    true
            );

            frog.setHealth(0.0);
            removeFrogBucket(player, itemInHand, false);
            event.setCancelled(true);
        }
    }

    private void removeFrogBucket(Player player, ItemStack stack, boolean giveEmptyBucket) {
        if (player.getGameMode() == GameMode.CREATIVE) return;
        if (stack.getAmount() > 1) {
            player.getInventory().setItemInMainHand(stack.subtract(1));
        } else {
            player.getInventory().setItemInMainHand(null);
        }

        if (giveEmptyBucket) player.getInventory().addItem(new ItemStack(Material.BUCKET));
    }

    private boolean isFrogInBucket(@NotNull ItemStack stack) {

        var itemMeta = stack.getItemMeta();
        if (itemMeta == null) return false;

        var container = itemMeta.getPersistentDataContainer();

        if (!container.has(frogNamespacedKey)) return false;
        var value = container.get(frogNamespacedKey, PersistentDataType.BOOLEAN);
        if (value == null) return false;

        return value;
    }

    private boolean deserializeFrogData(ItemStack stack, Location location) {

        var frog = location.getWorld().spawn(location, Frog.class);
        if (!frog.isInWorld()) return false;

        var itemMeta = stack.getItemMeta();
        if (itemMeta == null) return false;

        var container = itemMeta.getPersistentDataContainer();
        var customName = container.get(frogCustomNameKey, PersistentDataType.STRING);
        var variant = container.get(frogVariantKey, PersistentDataType.STRING);

        if (customName != null) {
            frog.customName(MiniMessage.miniMessage().deserialize(customName));
        }

        frog.setVariant(frogVariant(variant));
        return true;
    }

    private ItemStack createFrogItemStack(int amount, Frog.Variant variant, Component customName) {

        var itemStack = new ItemStack(Material.BUCKET, amount);
        var itemMeta = itemStack.getItemMeta();

        itemMeta.displayName(MiniMessage.miniMessage().deserialize(FROG_BUCKET_NAME));
        itemMeta.addEnchant(Enchantment.EFFICIENCY, 1, false);
        itemMeta.lore(List.of(Component.text(frogVariantName(variant))));

        var container = itemMeta.getPersistentDataContainer();
        container.set(frogCustomNameKey, PersistentDataType.STRING, customName != null ? MiniMessage.miniMessage().serialize(customName) : "");
        container.set(frogNamespacedKey, PersistentDataType.BOOLEAN, true);
        container.set(frogVariantKey, PersistentDataType.STRING, frogVariantName(variant));
        itemStack.setItemMeta(itemMeta);

        return itemStack;
    }

    private void addFrogBucketToPlayer(Player player, int amount, Component customName, Frog.Variant variant, boolean actionBarMessage) {

        var slot = player.getInventory().firstEmpty();
        if(slot == -1) {
            player.sendMessage(Component.translatable("plugin.inventory-full").arguments(plugin.getPrefix()));
            return;
        }

        var itemStack = createFrogItemStack(amount, variant, customName);
        player.getInventory().setItem(slot, itemStack);
        var message = Component.translatable("frog-bucket-added-to-inventory").arguments(
                plugin.getPrefix(),
                Component.text(FROG_BUCKET_NAME),
                Component.text(amount),
                player.displayName()
        );

        if (!actionBarMessage) {
            player.sendMessage(message);
        } else {
            player.sendActionBar(message);
        }
    }

    @NotNull
    private Frog.Variant frogVariant(@Nullable String variant) {

        if (variant == null) return DEFAULT_VARIANT;

        var key = NamespacedKey.fromString(variant.toLowerCase());
        if (key == null) return Frog.Variant.TEMPERATE;

        var frogVariant = Registry.FROG_VARIANT.get(key);

        return frogVariant != null ? frogVariant : DEFAULT_VARIANT;
    }

    private String frogVariantName(Frog.Variant variant) {
        return variant.key().value();
    }

    @Suggestions(value = "frogVariants")
    public List<String> frogVariants(CommandContext<CommandSender> context, String input) {
        var variants = Registry.FROG_VARIANT.stream()
                .map(Frog.Variant::getKey)
                .map(NamespacedKey::toString)
                .filter(key -> key.startsWith(input.toLowerCase()))
                .toList();
        return StringUtil.copyPartialMatches(input, variants, variants);
    }
}

File: src/main/kotlin/net/onelitefeather/stardust/service/SyncFrogService.kt
package net.onelitefeather.stardust.service

import io.papermc.paper.event.player.PrePlayerAttackEntityEvent
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import net.onelitefeather.stardust.StardustPlugin
import net.onelitefeather.stardust.util.PlayerUtils
import net.onelitefeather.stardust.util.StringUtils
import org.bukkit.*
import org.bukkit.command.CommandSender
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Frog
import org.bukkit.entity.Frog.Variant
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import org.bukkit.util.StringUtil
import org.incendo.cloud.annotation.specifier.Quoted
import org.incendo.cloud.annotations.Argument
import org.incendo.cloud.annotations.Command
import org.incendo.cloud.annotations.CommandDescription
import org.incendo.cloud.annotations.Permission
import org.incendo.cloud.annotations.suggestion.Suggestions
import org.incendo.cloud.context.CommandContext
import org.incendo.cloud.context.CommandInput

class SyncFrogService(val stardustPlugin: StardustPlugin) : Listener, StringUtils, PlayerUtils {

    private var frogNameSpacedKey: NamespacedKey = NamespacedKey(stardustPlugin, "frog_data_key")
    private var frogVariantKey: NamespacedKey = NamespacedKey(stardustPlugin, "frog_variant_key")
    private var frogCustomNameKey: NamespacedKey = NamespacedKey(stardustPlugin, "frog_custom_name_key")
    private var frogInBucket: Int = 1
    private var frogBucketName: String = "Frog Bucket"

    init {
        stardustPlugin.server.pluginManager.registerEvents(this, stardustPlugin)
    }

    @Command("frogbucket <customName> <variant> <amount>")
    @CommandDescription("Gives you a Frog bucket with the given amount!")
    @Permission("stardust.command.frogbucket")
    fun executeCommand(
        player: Player,
        @Quoted @Argument(value = "customName") customName: String,
        @Argument(value = "variant", suggestions = "frogVariants") variant: String,
        @Argument(value = "amount") amount: Int
    ) {
        addFrogBucketToPlayer(player, amount, MiniMessage.miniMessage().deserialize(customName), frogVariant(variant), false)
    }

    @EventHandler
    fun handlePlayerInteract(event: PlayerInteractEvent) {

        val player = event.player
        val itemStack = event.item ?: return
        val clickedBlock = event.clickedBlock ?: return

        if (event.action != Action.RIGHT_CLICK_BLOCK) return

        if (!isFrogInBucket(itemStack)) return
        if (deserializeFrogData(itemStack, clickedBlock.location.add(0.0, 1.0, 0.0))) {
            removeFrogBucket(player, player.inventory.itemInMainHand, true)
            player.sendMessage(
                Component.translatable("frog-bucket-spawn-success").arguments(stardustPlugin.getPluginPrefix())
            )
        } else {
            player.sendMessage(Component.translatable("frog-cannot-be-spawned").arguments(stardustPlugin.getPluginPrefix()))
        }
    }

    private fun removeFrogBucket(player: Player, itemStack: ItemStack, giveEmptyBucket: Boolean) {
        if (player.gameMode == GameMode.CREATIVE) return
        if (itemStack.amount > 1) {
            player.inventory.setItemInMainHand(itemStack.subtract(1))
        } else {
            player.inventory.setItemInMainHand(null)
        }

        if (giveEmptyBucket) player.inventory.addItem(ItemStack(Material.BUCKET))
    }

    @EventHandler
    fun handlePlayerEntityInteract(event: PrePlayerAttackEntityEvent) {

        val player = event.player
        val entity = event.attacked
        val itemInHand = player.inventory.itemInMainHand

        if (isFrogInBucket(itemInHand)) return
        if (entity is Frog && itemInHand.type == Material.BUCKET) {
            addFrogBucketToPlayer(
                player,
                1,
                entity.customName() ?: MiniMessage.miniMessage().deserialize(entity.name),
                entity.variant,
                true
            )
            entity.health = 0.0
            removeFrogBucket(player, itemInHand, false)
            event.isCancelled = true
        }
    }

    private fun isFrogInBucket(itemStack: ItemStack): Boolean {

        val itemMeta = itemStack.itemMeta ?: return false
        val container = itemMeta.persistentDataContainer

        if (!container.has(frogNameSpacedKey)) return false
        val value = container[frogNameSpacedKey, PersistentDataType.INTEGER]
        return value != null && value == frogInBucket
    }

    private fun deserializeFrogData(itemStack: ItemStack, location: Location): Boolean {

        val frog = location.world.spawn(location, Frog::class.java)
        if (!frog.isInWorld) return false
        val itemMeta = itemStack.itemMeta
        val container = itemMeta.persistentDataContainer
        if (!container.has(frogCustomNameKey) || !container.has(frogVariantKey)) return false

        val customName = container[frogCustomNameKey, PersistentDataType.STRING]
        val variant = container[frogVariantKey, PersistentDataType.STRING] ?: frogVariantName(Variant.TEMPERATE)

        if (customName != null) {
            frog.customName(colorText(customName))
        }

        frog.variant = frogVariant(variant)
        return true
    }

    private fun createFrogItemStack(amount: Int, variant: Variant, customName: Component?): ItemStack {

        val itemStack = ItemStack(Material.BUCKET, amount)
        val itemMeta = itemStack.itemMeta

        itemMeta.displayName(MiniMessage.miniMessage().deserialize(frogBucketName))

        itemMeta.addEnchant(Enchantment.EFFICIENCY, 1, false)
        itemMeta.addItemFlags(ItemFlag.HIDE_ENCHANTS)
        itemMeta.lore(listOf(MiniMessage.miniMessage().deserialize(frogVariantName(variant))))

        val container = itemMeta.persistentDataContainer

        if (customName != null) {
            container[frogCustomNameKey, PersistentDataType.STRING] = convertComponentToString(customName)
        }
        container[frogNameSpacedKey, PersistentDataType.INTEGER] = frogInBucket
        container[frogVariantKey, PersistentDataType.STRING] = variant.key().value()

        itemStack.itemMeta = itemMeta
        return itemStack
    }

    private fun addFrogBucketToPlayer(
        player: Player,
        amount: Int,
        customName: Component,
        variant: Variant,
        actionBarMessage: Boolean
    ) {

        val slot = player.inventory.firstEmpty()

        //If the slot returns -1 the inventory of the player is full
        if (slot == -1) {
            player.sendMessage(
                Component.translatable("plugin.inventory-full").arguments(stardustPlugin.getPluginPrefix())
            )
            return
        }

        val itemStack = createFrogItemStack(amount, variant, customName)
        player.inventory.setItem(slot, itemStack)
        val message = Component.translatable("frog-bucket-added-to-inventory").arguments(
            stardustPlugin.getPluginPrefix(),
            Component.text(frogBucketName),
            Component.text(amount),
            player.displayName()
        )

        if (!actionBarMessage) {
            player.sendMessage(message)
        } else {
            player.sendActionBar(message)
        }
    }

    private fun frogVariantName(variant: Variant): String = variant.key().value()

    private fun frogVariant(key: String): Variant {
        return Registry.FROG_VARIANT[NamespacedKey.fromString(key.lowercase())!!] as Variant
    }

    @Suggestions(value = "frogVariants")
    fun frogVariants(context: CommandContext<CommandSender>, input: String): MutableList<String> {
        val variants = Registry.FROG_VARIANT.stream().map { it.key.value() }.toList()
        return StringUtil.copyPartialMatches(input, variants, variants)
    }
}